README.md

```diff
@@ -175,18 +175,26 @@ The result of executing a GraphQL operation may contain partial data as well as
 Errors that happen during execution of the GraphQL operation typically become part of the result,
 as long as the server is still able to produce a well-formed response.
 
-In case of errors that completely prevent the successful execution of the request, the server SHOULD respond
-with the appropriate status code depending on the concrete error condition. For example:
+In case of errors that completely prevent the successful execution of the request, the server MUST respond
+with the appropriate status code depending on the concrete error condition.
 
-- `400` (Bad Request), in case of a validation error
-- `401` (Unauthorized), if an unauthenticated client is not allowed to access the schema
-- `500` (Internal Server Error), for unexpected failures in the server
+There are four types of validation errors:
 
-As an example, the way authorization errors may be handled illustrates the different error conditions.
-If a client is forbidden from making any GraphQL requests at all, the server may respond with
-status code `403` (Forbidden) and not return any GraphQL data.
-In contrast to that, a client might be allowed to access the schema, but requests some restricted fields.
-Given the server can provide at least partial `data`, the response would have status code `200` (OK) and 
-include `errors` as part of the GraphQL response.
+1. Unparseable or invalid request body, for example `NONSENSE` or `{"qeury": "{__typename}"}`
+2. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
+3. Custom document validations that can be performed before execution (e.g. graphql-depth-limit)
+4. Runtime validations performed by resolvers (i.e. during the execution of the GraphQL operation).
 
-It is RECOMMENDED to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+The error types 1. and 2. prevent execution of the request entirely and MUST result in
+status code `400` (Bad Request). Types 3. and 4. may still allow partial execution and
+MUST result in status code `200`.
+
+If an unauthenticated client is not allowed to access the schema, the server MUST respond with
+status code `401` (Unauthorized).
+
+If the server failed unexpectedly, it MUST respond with status code `500` (Internal Server Error).
+
+If a client is not authorized to access any of the requested fields, the server MUST respond with
+status code `403` (Forbidden) and not return any `data`.
```

Comment:
Requiring that we check that each individual field requested was forbidden is quite a burden, and those constraints live in the GraphQL schema itself rather than the GraphQL server, so enforcing them will be challenging (e.g. determining the difference between a field failing due to it being forbidden, versus it failing just due to an error occurring). Instead, 403 should be used when access to the GraphQL schema itself is denied.

I've also added an extra paragraph to deal with the two `SHOULD`s.

```suggestion
If a client is not authorized to access the GraphQL schema, the server SHOULD respond with
status code `403` (Forbidden) and not return any `data`.

If a client is not permitted to access the GraphQL schema for any reason, the server MUST respond with a `4xx` or `5xx` status code (a `2xx` status code MUST NOT be used).
```

Reply:
Suppose a server implements field level authorization, and the client accesses some allowed fields and some disallowed fields. The responses might contain partial `data`, as well as a bunch of `errors`. That MUST definitely be a 200.

However, when all requested fields are forbidden, the server is unable to produce any `data`. Consider the above section:

> In case of errors that completely prevent the successful execution of the request, the server MUST respond
with the appropriate status code depending on the concrete error condition.

The requests was not successful, the client did not get any of the data they asked for. I think that should be reflected in a non-200 status code.

> enforcing them will be challenging

Agreed, no one said this was going to be easy ;) In all seriousness, it seems doable. A well-built GraphQL server should be able to make it happen. If not, well, then it is fine if they are non-compliant with this aspect of the spec.

Reply:
I think that the spec should be written in a way that makes compliance easy - rather than talking about it in terms of "allowed fields and disallowed fields" which is not a concept that exists in the GraphQL spec, we should talk about it in terms of successful execution. In this case what you're really saying is "should resolving the request with graphql result in the data field being null, a 4xx or 5xx status should be used".

It's my belief that you should be able to build a GraphQL server without needing to know about the internal logic of the underlying GraphQL schema - everything the server needs to know should be available as a result of calling the `parse`, `validate` and `execute` GraphQL functions in whatever language you're using; these functions do not have any concept of "field-level authorization".

Reply:
I agree.  The server (the part that is outside of the GraphQL implementation) should not know or care about any part of the schema or any rules associated with it.  Though the server is permitted to determine whether a user is authorized to access the schema, and when not MUST respond with a 403 status code.  Equally, if the server decides that a user must be authenticated in order to access the schema, and an unauthenticated user makes a request, it MUST respond with a 401 status code.

If the GraphQL implementation successfully executed a request (even if all requested data is protected and the requesting user has no access to any of it), the server MUST respond with a 200 status code.